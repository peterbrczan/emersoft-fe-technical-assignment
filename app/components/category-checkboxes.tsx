'use client';

import type { FC } from 'react';
import React, { useEffect, useState } from 'react';
import { Category } from '@/app/models/types/category';
import { useTranslations } from 'use-intl';

export type CategoryCheckboxesProps = {
  categories: Category[];
  handleCategorySlugs: (categorySlugs: string[]) => void;
};

export const CategoryCheckboxes: FC<CategoryCheckboxesProps> = (props) => {
  const [categorySlugs, setCategorySlugs] = useState<string[]>([]);
  const textCategory = useTranslations('Category');

  useEffect(() => {
    props.handleCategorySlugs(categorySlugs);
  }, [categorySlugs]);

  function toggleCheckbox(categorySlug: string) {
    let copyCategorySlugs = [...categorySlugs];

    if (categorySlugs.includes(categorySlug)) {
      copyCategorySlugs = copyCategorySlugs.filter((copyCategorySlug) => copyCategorySlug !== categorySlug);
    } else {
      copyCategorySlugs.push(categorySlug);
    }

    setCategorySlugs(copyCategorySlugs);
  }

  return (
    <div className="flex gap-4 flex-wrap">
      {props.categories.map((category) => (
        <div key={category.slug} className="flex items-center">
          <input
            id={category.slug}
            type="checkbox"
            checked={categorySlugs.includes(category.slug)}
            className="w-[16px] h-[16px] rounded-[4px] cursor-pointer"
            onChange={() => toggleCheckbox(category.slug)}
          />
          <label for={category.slug} className="ml-[8px] text-sm font-medium text-gray-900 cursor-pointer">
            {textCategory(category.name)}
          </label>
        </div>
      ))}
    </div>
  );
};
