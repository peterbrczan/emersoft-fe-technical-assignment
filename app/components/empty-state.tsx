import type { FC, PropsWithChildren } from 'react';
import React from 'react';

export const EmptyState: FC<PropsWithChildren> = ({ children }) => {
  return <div className="bg-gray-200 h-32 p-2 flex items-center justify-center rounded-[4px]">{children}</div>;
};
