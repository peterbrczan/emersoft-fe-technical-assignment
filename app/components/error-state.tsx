import type { FC, PropsWithChildren } from 'react';
import React from 'react';

export const ErrorState: FC<PropsWithChildren> = ({ children }) => {
  return <div className="bg-red-300 h-32 p-2 flex items-center justify-center rounded-[4px]">{children}</div>;
};
