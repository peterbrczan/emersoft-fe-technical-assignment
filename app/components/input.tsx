'use client';

import type { FC } from 'react';
import React, { useEffect, useState } from 'react';
import { useTranslations } from 'use-intl';
import { useDebounce } from 'use-debounce';

export type InputProps = {
  handleSearchValue: (value: string) => void;
};

export const Input: FC<InputProps> = (props) => {
  const [value, setValue] = useState('');
  const [searchValue] = useDebounce(value, 500);

  useEffect(() => {
    props.handleSearchValue(searchValue);
  }, [searchValue]);

  const textCommon = useTranslations('Common');

  return (
    <div className="relative">
      <div className="absolute inset-y-0 left-0 flex items-center pl-3 pointer-events-none">
        <svg className="w-4 h-4 text-gray-500" fill="none" viewBox="0 0 20 20">
          <path
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            d="m19 19-4-4m0-7A7 7 0 1 1 1 8a7 7 0 0 1 14 0Z"
          />
        </svg>
      </div>
      <input
        type="search"
        className="block w-full p-2 pl-8 text-sm text-gray-900 border border-gray-300 rounded-[4px] bg-gray-100 "
        placeholder={textCommon('Search by a title')}
        onChange={({ currentTarget: { value } }) => setValue(value)}
        data-testid="search-input"
      />
    </div>
  );
};
