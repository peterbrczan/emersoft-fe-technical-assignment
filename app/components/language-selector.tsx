'use client';

import type { FC } from 'react';
import React, { useMemo } from 'react';
import { useLocale } from 'use-intl';
import { Locale } from '@/app/models/enums/locale';
import Link from 'next/link';
import { usePathname } from 'next/navigation';

export const LanguageSelector: FC = () => {
  const locale = useLocale() as Locale;
  const pathname = usePathname();

  const flag = useMemo(() => {
    switch (locale) {
      case Locale.HU: {
        return { src: 'https://flagcdn.com/w40/hu.png', alt: 'Hungarian Flag' };
      }
      case Locale.EN:
      default: {
        return {
          src: 'https://flagcdn.com/w40/gb.png',
          alt: 'Great Britain flag',
        };
      }
    }
  }, [locale]);

  function getNextHref() {
    const nextUrl = pathname;
    if (nextUrl.startsWith(`/${Locale.HU}`)) {
      return nextUrl.replace(`/${Locale.HU}`, `/${Locale.EN}`);
    } else if (nextUrl.startsWith(`/${Locale.EN}`)) {
      return nextUrl.replace(`/${Locale.EN}`, `/${Locale.HU}`);
    } else {
      return `${Locale.HU}${nextUrl}`;
    }
  }

  return (
    <Link href={getNextHref()} className="w-[20px] h-[20px] rounded-full">
      <img className="w-[20px] h-[20px] rounded-full" src={flag.src} alt={flag.alt} />
    </Link>
  );
};
