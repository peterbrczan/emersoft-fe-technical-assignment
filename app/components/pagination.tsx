'use client';

import type { FC } from 'react';
import React, { useEffect, useState } from 'react';
import { useTranslations } from 'use-intl';

export type PaginationProps = {
  pageNumber: number;
  numberOfTotalPages: number;
  handlePageNumber: (pageNumber: number) => void;
};

export const Pagination: FC<PaginationProps> = (props) => {
  const [pageNumber, setPageNumber] = useState(props.pageNumber);
  const textLabel = useTranslations('Label');

  useEffect(() => {
    props.handlePageNumber(pageNumber);
  }, [pageNumber]);

  const isNextButtonDisabled = pageNumber + 1 > props.numberOfTotalPages;
  const isPreviousButtonDisabled = pageNumber - 1 < 1;

  function handleNextPage() {
    if (!isNextButtonDisabled) {
      setPageNumber((previousPageNumber) => previousPageNumber + 1);
    }
  }

  function handlePreviousPage() {
    if (!isPreviousButtonDisabled) {
      setPageNumber((previousPageNumber) => previousPageNumber - 1);
    }
  }

  return (
    <div className="grid grid-cols-2 gap-1">
      <button
        className="bg-pink-700 hover:bg-pink-500 text-white font-bold py-2 px-4 border border-pink-700 rounded disabled:bg-pink-200 disabled:border-0"
        onClick={handlePreviousPage}
        disabled={isPreviousButtonDisabled}
      >
        <p className="text-sm">{textLabel('Previous')}</p>
      </button>
      <button
        className="bg-pink-700 hover:bg-pink-500 text-white font-bold py-2 px-4 border border-pink-700 rounded disabled:bg-pink-200 disabled:border-0"
        onClick={handleNextPage}
        disabled={isNextButtonDisabled}
      >
        <p className="text-sm">{textLabel('Next')}</p>
      </button>
    </div>
  );
};
