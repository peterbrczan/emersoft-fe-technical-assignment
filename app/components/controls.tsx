import type { FC } from 'react';
import React, { useState } from 'react';
import { Input } from '@/app/components/input';
import { useGetBlog } from '@/app/hooks/use-get-blog';

export const Controls: FC = (props) => {
  const [searchValue, setSearchValue] = useState('');
  const getBlog = useGetBlog({ searchValue });

  return (
    <div>
      <Input handleSearchValue={setSearchValue} isLoading={getBlog.isLoading} />
    </div>
  );
};
