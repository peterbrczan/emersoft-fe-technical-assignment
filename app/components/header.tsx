'use client';

import type { FC } from 'react';
import React from 'react';
import { useTranslations } from 'use-intl';
import { LanguageSelector } from '@/app/components/language-selector';

export const Header: FC = () => {
  const textCommon = useTranslations('Common');

  return (
    <div className="border-b-2 border-gray-100 min-h-[80px] grid grid-cols-[1fr_auto] justify-between p-4 gap-[4px]">
      <div className="self-center flex flex-wrap gap-2 items-center">
        <h1 className="text-2xl font-extrabold">Emersoft</h1>
        <div className="w-[8px] h-[8px] rounded-full bg-pink-700"></div>
        <p>{textCommon('Frontend Technical Assignment')}</p>
      </div>
      <div className="justify-self-end self-center">
        <LanguageSelector />
      </div>
    </div>
  );
};
