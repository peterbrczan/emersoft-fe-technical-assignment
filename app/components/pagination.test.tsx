import { render } from '@/app/utils/test-renderer';
import { Pagination, PaginationProps } from '@/app/components/pagination';

describe('[component]: pagination', () => {
  let defaultProps: PaginationProps;

  beforeEach(() => {
    defaultProps = {
      pageNumber: 1,
      numberOfTotalPages: 4,
      handlePageNumber: jest.fn(),
    };
  });

  it('should match snapshot', () => {
    const { asFragment } = render(<Pagination {...defaultProps} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('should with disabled previous button match snapshot', () => {
    defaultProps.numberOfTotalPages = 1;
    const { asFragment } = render(<Pagination {...defaultProps} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('should with disabled next button match snapshot', () => {
    defaultProps.numberOfTotalPages = 4;
    const { asFragment } = render(<Pagination {...defaultProps} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
