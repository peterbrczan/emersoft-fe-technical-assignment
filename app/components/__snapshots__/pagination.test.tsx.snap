// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[component]: pagination should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid grid-cols-2 gap-1"
  >
    <button
      class="bg-pink-700 hover:bg-pink-500 text-white font-bold py-2 px-4 border border-pink-700 rounded disabled:bg-pink-200 disabled:border-0"
      disabled=""
    >
      <p
        class="text-sm"
      >
        Previous
      </p>
    </button>
    <button
      class="bg-pink-700 hover:bg-pink-500 text-white font-bold py-2 px-4 border border-pink-700 rounded disabled:bg-pink-200 disabled:border-0"
    >
      <p
        class="text-sm"
      >
        Next
      </p>
    </button>
  </div>
</DocumentFragment>
`;

exports[`[component]: pagination should with disabled next button match snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid grid-cols-2 gap-1"
  >
    <button
      class="bg-pink-700 hover:bg-pink-500 text-white font-bold py-2 px-4 border border-pink-700 rounded disabled:bg-pink-200 disabled:border-0"
      disabled=""
    >
      <p
        class="text-sm"
      >
        Previous
      </p>
    </button>
    <button
      class="bg-pink-700 hover:bg-pink-500 text-white font-bold py-2 px-4 border border-pink-700 rounded disabled:bg-pink-200 disabled:border-0"
    >
      <p
        class="text-sm"
      >
        Next
      </p>
    </button>
  </div>
</DocumentFragment>
`;

exports[`[component]: pagination should with disabled previous button match snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid grid-cols-2 gap-1"
  >
    <button
      class="bg-pink-700 hover:bg-pink-500 text-white font-bold py-2 px-4 border border-pink-700 rounded disabled:bg-pink-200 disabled:border-0"
      disabled=""
    >
      <p
        class="text-sm"
      >
        Previous
      </p>
    </button>
    <button
      class="bg-pink-700 hover:bg-pink-500 text-white font-bold py-2 px-4 border border-pink-700 rounded disabled:bg-pink-200 disabled:border-0"
      disabled=""
    >
      <p
        class="text-sm"
      >
        Next
      </p>
    </button>
  </div>
</DocumentFragment>
`;
